File: hooks/python-hooks/lambda-function-invoker/example-lambda-functions/s3-versioning-enabled/golang/main.go
// Example Lambda function called by AWSSamples::LambdaFunctionInvoker::Hook.

package main

import (
	"context"
	"fmt"
	"github.com/aws/aws-lambda-go/lambda"
	"log"
)

// If you choose to invoke the hook for all the invocation points
// (default behavior), and to not set `TargetFilters` for your hook,
// you can define, in the example Lambda function code, for which
// invocation points to run the code in this function. It is important
// to understand that if you choose this method as opposed to target
// filters, the hook will be invoked anyway for all of the AWS
// resource types in your templates across all the invocation points,
// and you'll be billed for all the invocations. To use this feature,
// comment out the line for `hookInvocationPoints` that corresponds to
// the invocation point you do not want to run. Later on in this
// example code, you'll find a conditional block that will return
// SUCCESS if the current invocation point is something you do not
// need to evaluate on during the current invocation.
var hookInvocationPoints = []string{
	"CREATE_PRE_PROVISION",
	"UPDATE_PRE_PROVISION",
	"DELETE_PRE_PROVISION",
}

// Add AWS resource type target that this code will evaluate. This
// example function validates if versioning is enabled for an S3
// bucket; hence, add AWS::S3::Bucket here.
var targetNames = []string{
	"AWS::S3::Bucket",
}

type InputEvent struct {
	Request map[string]interface{} `json:"request"`
}

func HandleRequest(ctx context.Context, event *InputEvent) (*map[string]interface{}, error) {

	// Get the entire request from the input event.
	request := event.Request

	hookContext := request["hookContext"].(map[string]interface{})

	// Get the invocation point for the hook that is calling this
	// function; for example: `CREATE_PRE_PROVISION`.
	invocationPoint := hookContext["invocationPoint"].(string)
	log.Printf("Invocation point: %s", invocationPoint)

	targetName := hookContext["targetName"].(string)
	log.Printf("Target name: %s", targetName)

	// If you're not using TargetFilters for the hook that invokes
	// this function, return SUCCESS immediately if you do not
	// want to perform policy-as-code validation for a given hook
	// invocation point that is commented out in
	// hookInvocationPoints at the beginning of this code, or for
	// target names you've not added to targetNames above.
	if !stringsSliceHas(hookInvocationPoints, invocationPoint) || !stringsSliceHas(targetNames, targetName) {
		message := fmt.Sprintf("Skipping %s evaluation for %s.", targetName, invocationPoint)
		log.Printf(message)
		response := map[string]interface{}{
			"status":               "SUCCESS",
			"errorCode":            nil,
			"message":              message,
			"callbackContext":      nil,
			"callbackDelaySeconds": 0,
		}
		return &response, nil
	}

	// Get the model of the resource type target.
	targetModel := hookContext["targetModel"].(map[string]interface{})

	resourceProperties := make(map[string]interface{})

	// Get the resource properties of the target resource type;
	// this should include the properties for the target resource
	// type that the user specified in their CloudFormation
	// template.
	if resourcePropertiesValue, hasResourceProperties := targetModel["resourceProperties"]; hasResourceProperties {
		resourceProperties = resourcePropertiesValue.(map[string]interface{})
	}
	// log.Printf("%s", resourceProperties)

	// The code below needs to validate not only if `Status` for
	// `VersioningConfiguration` is set to `Enabled`, but also if
	// all the other parent properties are present in the
	// template.
	versioningConfiguration := make(map[string]interface{})
	if versioningConfigurationValue, hasVersioningConfiguration := resourceProperties["VersioningConfiguration"]; hasVersioningConfiguration {
		versioningConfiguration = versioningConfigurationValue.(map[string]interface{})
	}
	// log.Printf("%s", versioningConfiguration)

	versioningConfigurationStatus := ""
	if versioningConfigurationStatusValue, hasVersioningConfigurationStatus := versioningConfiguration["Status"]; hasVersioningConfigurationStatus {
		versioningConfigurationStatus = versioningConfigurationStatusValue.(string)
	}
	// log.Printf("%s", versioningConfigurationStatus)

	if versioningConfigurationStatus != "Enabled" {
		message := "Versioning not set or not enabled for the S3 bucket."
		log.Printf(message)
		// Return FAILED to the hook, using a payload format
		// that the hook expects: a data structure that
		// contains information on status, error code,
		// message, callback context, and callback delay in
		// seconds. This format is modeled after a
		// `ProgressEvent` object:
		// https://docs.aws.amazon.com/cloudformation-cli/latest/userguide/resource-type-test-progressevent.html
		response := map[string]interface{}{
			"status":               "FAILED",
			"errorCode":            "NonCompliant",
			"message":              message,
			"callbackContext":      nil,
			"callbackDelaySeconds": 0,
		}
		return &response, nil
	}

	// Return success if the resource is compliant.
	message := "Versioning is enabled for the S3 bucket."
	log.Printf(message)
	response := map[string]interface{}{
		"status":               "SUCCESS",
		"errorCode":            nil,
		"message":              message,
		"callbackContext":      nil,
		"callbackDelaySeconds": 0,
	}
	return &response, nil
}

func stringsSliceHas(stringsSlice []string, inputString string) bool {
	for _, stringInSlice := range stringsSlice {
		if stringInSlice == inputString {
			return true
		}
	}
	return false
}

func main() {
	lambda.Start(HandleRequest)
}

File: hooks/python-hooks/lambda-function-invoker/example-lambda-functions/s3-versioning-enabled/cpp/main.cpp
// Example Lambda function called by AWSSamples::LambdaFunctionInvoker::Hook.

#include <exception>
#include <iostream>
#include <set>
#include <stdexcept>
#include <string>

#include <aws/core/utils/json/JsonSerializer.h>
#include <aws/lambda-runtime/runtime.h>

using namespace std;
using namespace Aws::Utils::Json;
using namespace aws::lambda_runtime;

// If you choose to invoke the hook for all the invocation points
// (default behavior), and to not set `TargetFilters` for your hook,
// you can define, in the example Lambda function code, for which
// invocation points to run the code in this function. It is important
// to understand that if you choose this method as opposed to target
// filters, the hook will be invoked anyway for all of the AWS
// resource types in your templates across all the invocation points,
// and you'll be billed for all the invocations. To use this feature,
// comment out the line for `hookInvocationPoints` that corresponds to
// the invocation point you do not want to run. Later on in this
// example code, you'll find a conditional block that will return
// SUCCESS if the current invocation point is something you do not
// need to evaluate on during the current invocation.
set<string> hookInvocationPoints{
    "CREATE_PRE_PROVISION",
    "UPDATE_PRE_PROVISION",
    "DELETE_PRE_PROVISION",
};

// Add AWS resource type target that this code will evaluate. This
// example function validates if versioning is enabled for an S3
// bucket; hence, add AWS::S3::Bucket here.
set<string> targetNames{
    "AWS::S3::Bucket",
};

// MIME type to return in invocation responses.
const string invocationResponseMimeType = "application/json";

invocation_response handler(invocation_request const &request) {
  JsonValue payload(request.payload);
  JsonView payloadView = payload.View();

  try {
    // Get the entire request from the input event.
    JsonView hookRequest = payloadView.GetObject("request");
    JsonView hookContext = hookRequest.GetObject("hookContext");

    // Get the invocation point for the hook that is calling this
    // function; for example: `CREATE_PRE_PROVISION`.
    string invocationPoint = hookContext.GetString("invocationPoint");
    cout << "Invocation point: " << invocationPoint << endl;

    // Get the name of the resource type target that invoked the hook.
    // For example: `AWS::S3::Bucket`.
    string targetName = hookContext.GetString("targetName");
    cout << "Target name: " << targetName << endl;

    // If you're not using TargetFilters for the hook that invokes
    // this function, return SUCCESS immediately if you do not want to
    // perform policy-as-code validation for a given hook invocation
    // point that is commented out in hookInvocationPoints at the
    // beginning of this code, or for target names you've not added to
    // targetNames above.
    if (hookInvocationPoints.find(invocationPoint) ==
            hookInvocationPoints.end() ||
        targetNames.find(targetName) == targetNames.end()) {
      JsonValue payload;
      string message =
          "Skipping " + targetName + " evaluation for " + invocationPoint + ".";
      cout << message << endl;
      payload.WithString("status", "SUCCESS");
      payload.WithString("errorCode", "");
      payload.WithString("message", message);
      payload.WithString("callbackContext", {});
      payload.WithInteger("callbackDelaySeconds", 0);
      auto response = payload.View().WriteReadable();
      // cout << response << endl;
      return invocation_response::success(response, invocationResponseMimeType);
    }

    // Get the model of the resource type target.
    auto targetModel = hookContext.GetObject("targetModel");

    // Get the resource properties of the target resource type; this
    // should include the properties for the target resource type that
    // the user specified in their CloudFormation template.
    auto resourceProperties = targetModel.GetObject("resourceProperties");

    auto versioningConfiguration =
        resourceProperties.GetObject("VersioningConfiguration");
    auto versioningConfigurationStatus =
        versioningConfiguration.GetString("Status");

    if (!resourceProperties.IsObject() || !versioningConfiguration.IsObject() ||
        !versioningConfiguration.KeyExists("Status") ||
        versioningConfigurationStatus != "Enabled") {
      JsonValue payload;
      string message = "Versioning not set or not enabled for the S3 bucket.";
      cerr << message << endl;
      // Return FAILED to the hook, using a payload format that the
      // hook expects: a data structure that contains information on
      // status, error code, message, callback context, and callback
      // delay in seconds. This format is modeled after a
      // `ProgressEvent` object:
      // https://docs.aws.amazon.com/cloudformation-cli/latest/userguide/resource-type-test-progressevent.html
      // The expected data structure is serialized with
      // Aws::Utils::Json::JsonValue::View() and WriteReadable() into
      // the invocation response.
      payload.WithString("status", "FAILED");
      payload.WithString("errorCode", "NonCompliant");
      payload.WithString("message", message);
      payload.WithString("callbackContext", {});
      payload.WithInteger("callbackDelaySeconds", 0);
      auto response = payload.View().WriteReadable();
      // cout << response << endl;
      return invocation_response::success(response, invocationResponseMimeType);
    }

    JsonValue payload;
    string message = "Versioning is enabled for the S3 bucket.";
    cout << message << endl;
    payload.WithString("status", "SUCCESS");
    payload.WithString("errorCode", "");
    payload.WithString("message", message);
    payload.WithString("callbackContext", {});
    payload.WithInteger("callbackDelaySeconds", 0);
    auto response = payload.View().WriteReadable();
    // cout << response << endl;
    return invocation_response::success(response, invocationResponseMimeType);
  } catch (const runtime_error &runtimeError) {
    cerr << "Runtime error: " << runtimeError.what() << endl;
    JsonValue payload;
    payload.WithString("status", "FAILED");
    payload.WithString("errorCode", "InternalFailure");
    payload.WithString("message", runtimeError.what());
    payload.WithString("callbackContext", {});
    payload.WithInteger("callbackDelaySeconds", 0);
    auto response = payload.View().WriteReadable();
    // cout << response << endl;
    return invocation_response::success(response, invocationResponseMimeType);
  } catch (const exception &exception) {
    cerr << "Exception: " << exception.what() << endl;
    JsonValue payload;
    payload.WithString("status", "FAILED");
    payload.WithString("errorCode", "InternalFailure");
    payload.WithString("message", exception.what());
    payload.WithString("callbackContext", {});
    payload.WithInteger("callbackDelaySeconds", 0);
    auto response = payload.View().WriteReadable();
    // cout << response << endl;
    return invocation_response::success(response, invocationResponseMimeType);
  } catch (...) {
    JsonValue payload;
    string message = "Unknown error occurred.";
    cerr << message << endl;
    payload.WithString("status", "FAILED");
    payload.WithString("errorCode", "InternalFailure");
    payload.WithString("message", message);
    payload.WithString("callbackContext", {});
    payload.WithInteger("callbackDelaySeconds", 0);
    auto response = payload.View().WriteReadable();
    // cout << response << endl;
    return invocation_response::success(response, invocationResponseMimeType);
  }
}

int main() {
  cout << "Running s3-versioning-enabled." << endl;
  run_handler(handler);
  return 0;
}
